golang/cosmos/x/vlocalchain/keeper/keeper.go

```diff
@@ -0,0 +1,263 @@
+package keeper
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/gogo/protobuf/jsonpb"
+	"github.com/gogo/protobuf/proto"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
+	storetypes "github.com/cosmos/cosmos-sdk/store/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkaddress "github.com/cosmos/cosmos-sdk/types/address"
+
+	abci "github.com/tendermint/tendermint/abci/types"
+
+	"github.com/Agoric/agoric-sdk/golang/cosmos/x/vlocalchain/types"
+)
+
+// Keeper maintains the link to data storage and exposes getter/setter methods for the various parts of the state machine
+type Keeper struct {
+	key         storetypes.StoreKey
+	cdc         codec.Codec
+	msgRouter   types.MsgRouter
+	queryRouter types.GRPCQueryRouter
+}
+
+// NewKeeper creates a new dIBC Keeper instance
+func NewKeeper(
+	cdc codec.Codec,
+	key storetypes.StoreKey,
+	msgRouter types.MsgRouter,
+	queryRouter types.GRPCQueryRouter,
+) Keeper {
+	return Keeper{
+		key:         key,
+		cdc:         cdc,
+		msgRouter:   msgRouter,
+		queryRouter: queryRouter,
+	}
+}
+
+// DeserializeRequests uses the global proto codec's proto3 JSON
+// unmarshaller and then unpacks the Any types into actual objects.
+func (k Keeper) DeserializeRequests(bz []byte) ([]types.QueryRequest, error) {
+	var cosmosTx types.CosmosTx
+	unmarshaler := &jsonpb.Unmarshaler{}
+
+	// Convert the bz byte slice to an io.Reader
+	rdr := bytes.NewReader(bz)
+	if err := unmarshaler.Unmarshal(rdr, &cosmosTx); err != nil {
+		return nil, err
+	}
+
+	msgs := make([]types.QueryRequest, len(cosmosTx.Messages))
+	for i, anyMsg := range cosmosTx.Messages {
+		// TODO: This is a hack that only works if the names are conventionally
```

Comment:
consider following the TODO immediately with what is to be done (so greps for "TODO" show it)